### ISSUE TITLE
[Feature request] AKS w/o SLB with implicit egress (no PIP, no UDR)

### ISSUE THREAD

**Issue by baboune** [Azure/AKS#1524]
**What happened**:
Deployed a Service using the "type: "LoadBalancer".  As soon as the IP was allocated to the service, the AKS cluster connectivity failed:
- Pods could not pull images from dockerhub
`Warning  Failed     27m (x4 over 30m)  kubelet, aks-c8m32-24599614-vmss000001  Failed to pull image "alpine": rpc error: code = Unknown desc = Error response from daemon: Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)`
- Nodes went into a "NotReady" state.

Remove the Service with LoadBalancer type, and the connectivity will be restored in a few minutes.

**What you expected to happen**:
The service to deploy and to get an IP address that could be reach from outside.

**How to reproduce it (as minimally and precisely as possible)**:
Create an AKS cluster with Azure Advanced CNI, Standard Load Balancer, and select the egress outbound type to UDR.

See https://docs.microsoft.com/en-us/azure/aks/egress-outboundtype

**Anything else we need to know?**:

Our deployment is similar to https://docs.microsoft.com/en-us/azure/aks/egress-outboundtype except that we do not use an Azure Firewall instance, we only have a route table associated with the VNET subnet used by AKS CNI.

We are in a situation where we use AKS Azure CNI with SLB UDR because we are forbidden by a subscription wide policy set by our IT department to create any public IPs.  In short, we can not use any Azure service that will instantiate a public IP.

As such:
- Azure SLB with public IP is forbidden
- Azure Firewall is forbidden
- etc.

Hence we create an AKS cluster with network set to azure CNI, the PoD CIDR is set to point a Peered VNET subnet range, and the SLB outbound type is set to UDR.  In order for AKS to deploy we must add a route table and associate it to the peered VNET subnet.

We have no Network Virtual Appliance nor Azure FW IPs to add in the routes.

Once the AKS instance is up and running, a typical workflow for this problem is as follow...

At 4 minutes the nodes are “ready”.

```
✗ k get nodes
NAME                            STATUS   ROLES   AGE     VERSION
aks-c8m32-24599614-vmss000000   Ready    agent   4m42s   v1.15.7
aks-c8m32-24599614-vmss000001   Ready    agent   4m10s   v1.15.7
 ```

I then deploy Prometheus:

`$ helm upgrade --install prometheus --namespace monitoring stable/prometheus
 
All pods pods start.
```
~ k -n monitoring get pods -w
NAME                                             READY   STATUS    RESTARTS   AGE
prometheus-alertmanager-64c9f45ff-zhrdw          2/2     Running   0          3m23s
prometheus-kube-state-metrics-5d95788848-ncl4c   1/1     Running   0          3m23s
prometheus-node-exporter-mqqlh                   1/1     Running   0          3m23s
prometheus-node-exporter-qtxwt                   1/1     Running   0          3m23s
prometheus-server-0                              2/2     Running   0          3m23s
prometheus-server-1                              2/2     Running   0          3m23s
 ```
My guess is everything works because at this point I am not using the Standard Load Balancer.
 
I can kubectl-node and get the shell:
 
```
$ ./kubectl-node_shell aks-c8m32-24599614-vmss000000
spawning "nsenter-4adhz2" on "aks-c8m32-24599614-vmss000000"
 
If you don't see a command prompt, try pressing enter.
 
root@aks-c8m32-24599614-vmss000000:/#
root@aks-c8m32-24599614-vmss000000:/#
```
 
Now I will deploy the ingress-controller, the ingress controller has a service manifest of type "LoadBalancer, and uses the Standard Load Balancer:
 
At this point, the nodes are still ready:
```
✗ k get nodes
NAME                            STATUS   ROLES   AGE   VERSION
aks-c8m32-24599614-vmss000000   Ready    agent   13m   v1.15.7
aks-c8m32-24599614-vmss000001   Ready    agent   12m   v1.15.7
 ```
SLB is now in use:
```
✗ k -n ingress-controller get svc
NAME                            TYPE           CLUSTER-IP     EXTERNAL-IP     PORT(S)                      AGE
nginx-ingress-controller        LoadBalancer   10.0.234.73    100.126.23.36   80:32351/TCP,443:30416/TCP   43s
nginx-ingress-default-backend   ClusterIP      10.0.126.106   <none>          80/TCP                       43s
 ```
Then I wait 4 min.
 
Try to deploy an “Alpine” pod:
```
kubernetes (develop) ✗ k run -it --rm --image alpine tt -- sh
kubectl run --generator=deployment/apps.v1 is DEPRECATED and will be removed in a future version. Use kubectl run --generator=run-pod/v1 or kubectl create instead.
 
```
And now it fails:
```
~ k get pods -w
➜  lmcnise@EMB-R1DXJG5J ~ k get pods -w
NAME                  READY   STATUS             RESTARTS   AGE
nsenter-4adhz2        1/1     Terminating        0          8m45s
tt-756fb699c6-5r247   0/1     ImagePullBackOff   0          2m33s
```

And both nodes are now “NotReady”:
```
kubernetes (develop) ✗ k get nodes
NAME                            STATUS     ROLES   AGE   VERSION
aks-c8m32-24599614-vmss000000   NotReady   agent   21m   v1.15.7
aks-c8m32-24599614-vmss000001   NotReady   agent   20m   v1.15.7
 ```

I remove the nginx-ingress that uses the Load Balancer:
`✗ helm -n ingress-controller uninstall nginx-ingress`
 
And my pod can now pull:
```
~ k get pods
NAME                  READY   STATUS    RESTARTS   AGE
tt-756fb699c6-5r247   1/1     Running   0          5m8s
```
 
Conclusion: The problem lies in the SLB.  As soon as it is in use, the cluster fails to have a normal network ie as soon as a Service uses the “LoadBalancer” type.  The issue must lie in the “route table” and its required routes.  What does the SLB expect there? And why?
 
I can observe the following in the nginx Service:
```
(develop) ✗ k -n ingress-controller get svc nginx-ingress-controller -o yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.beta.kubernetes.io/azure-load-balancer-internal: "true"
  creationTimestamp: "2020-03-24T09:53:59Z"
  labels:
    app: nginx-ingress
    chart: nginx-ingress-1.33.4
    component: controller
    heritage: Helm
    release: nginx-ingress
  name: nginx-ingress-controller
  namespace: ingress-controller
  resourceVersion: "9124"
  selfLink: /api/v1/namespaces/ingress-controller/services/nginx-ingress-controller
  uid: 339f8c66-4a63-4dc4-bb00-388341b9d0d1
spec:
  clusterIP: 10.0.162.143
  externalTrafficPolicy: Local
  healthCheckNodePort: 30494
  ports:
  - name: http
    nodePort: 31607
    port: 80
    protocol: TCP
    targetPort: http
  - name: https
    nodePort: 30729
    port: 443
    protocol: TCP
    targetPort: https
  selector:
    app: nginx-ingress
    component: controller
    release: nginx-ingress
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - ip: 100.126.23.36
```
Describe endpoints:

```
✗ k -n ingress-controller describe ep nginx-ingress-controller
Name:         nginx-ingress-controller
Namespace:    ingress-controller
Labels:       app=nginx-ingress
              chart=nginx-ingress-1.33.4
              component=controller
              heritage=Helm
              release=nginx-ingress
Annotations:  endpoints.kubernetes.io/last-change-trigger-time: 2020-03-24T10:00:30Z
Subsets:
  Addresses:          <none>
  NotReadyAddresses:  100.126.23.18,100.126.23.33
  Ports:
    Name   Port  Protocol
    ----   ----  --------
    http   80    TCP
    https  443   TCP

Events:  <none>
```

Describe service:
```
 ✗ k -n ingress-controller describe svc
Name:                     nginx-ingress-controller
Namespace:                ingress-controller
Labels:                   app=nginx-ingress
                          chart=nginx-ingress-1.33.4
                          component=controller
                          heritage=Helm
                          release=nginx-ingress
Annotations:              service.beta.kubernetes.io/azure-load-balancer-internal: true
Selector:                 app=nginx-ingress,component=controller,release=nginx-ingress
Type:                     LoadBalancer
IP:                       10.0.162.143
LoadBalancer Ingress:     100.126.23.36
Port:                     http  80/TCP
TargetPort:               http/TCP
NodePort:                 http  31607/TCP
Endpoints:
Port:                     https  443/TCP
TargetPort:               https/TCP
NodePort:                 https  30729/TCP
Endpoints:
Session Affinity:         None
External Traffic Policy:  Local
HealthCheck NodePort:     30494
Events:
  Type     Reason                   Age   From                Message
  ----     ------                   ----  ----                -------
  Normal   EnsuringLoadBalancer     34m   service-controller  Ensuring load balancer
  Normal   EnsuredLoadBalancer      34m   service-controller  Ensured load balancer
  Warning  UnAvailableLoadBalancer  28m   service-controller  There are no available nodes for LoadBalancer

Name:              nginx-ingress-default-backend
Namespace:         ingress-controller
Labels:            app=nginx-ingress
                   chart=nginx-ingress-1.33.4
                   component=default-backend
                   heritage=Helm
                   release=nginx-ingress
Annotations:       <none>
Selector:          app=nginx-ingress,component=default-backend,release=nginx-ingress
Type:              ClusterIP
IP:                10.0.199.46
Port:              http  80/TCP
TargetPort:        http/TCP
Endpoints:
Session Affinity:  None
Events:            <none>
```

The "**Warning  UnAvailableLoadBalancer  28m   service-controller  There are no available nodes for LoadBalancer**" is IMO a concern.

Any Service of type LoadBalancer will trigger the same failure.

What are the UDR outbound type dependencies?
What can we do to bypass this problem? Can we add routes?

**Environment**:
- Kubernetes version (use `kubectl version`): 
1.15.7
- Size of cluster (how many worker nodes are in the cluster?)
2 nodes
- General description of workloads in the cluster (e.g. HTTP microservices, Java app, Ruby on Rails, machine learning, etc.)
None, first pods can fail the cluster.
- Others:
Azure East US zone.

**Comment by gtracer** [comment]
@baboune, how have you setup your egress? With UDR (userDefinedRouting) all egress must be setup from the rules defined in your UDR. I am assuming you have a 0.0.0.0/0 route which redirects all traffic from SLB to whatever network appliance you are using for egress?

When you add the ingress controller (which adds a svc=loadBalancer which then adds a SLB rule), that should only impact inbound traffic. Egress should still be setup via your UDR. 

Can you confirm that egress traffic was indeed going through your UDR to your egress application in your peered vnet? I want to make sure it wasn't using Azure VM default networking (each VM has default egress until a SLB is configured). 

Confirming this will be specific to your egress setup but to give you a parallel, if you were using firewall you could check the packet metrics on firewall and make sure lots of packets are flying through it

**Comment by baboune** [comment]
@gtracer We have a Route Table with the  0.0.0.0/0 route.

We do not setup any network appliance for egress.  It is unclear why would that be needed?

> Can you confirm that egress traffic was indeed going through your UDR to your egress application in your peered vnet? I want to make sure it wasn't using Azure VM default networking (each VM has default egress until a SLB is configured).
What I observe is that all is fine until I deploy the svc=LoadBalancer. Also, the SLB assigned IP is reachable as ingress.  
How could I verify that egress is still going through the UDRs after the SLB is configured?

**Comment by gtracer** [comment]
@baboune the cluster has to get egress in some way (either SLB with public IP or UDR via you)

Here is how egress flows
1. If no load balancer is configured and no default routes taking egress traffic then you will get `azure VM default egress` which provides egress from the VM itself. This is not to be relied on as it changes when you get a load balancer. It appears your cluster is using this although that surprises me because if you setup your route table with the subnet then I would have expected the default `0.0.0.0/0` rule to be taking all egress traffic

2. With SLB, normally we setup a public IP on the load balancer and all traffic is directed through that IP for egress traffic.

3. When you use UDR then we don't create a public IP on the SLB. Here is what we expect in the UDR case: 
- A route table associated to your custom subnet
- That route table to direct all egress traffic `0.0.0.0/0` to your own egress device (we use a firewall in the docs but other personal devices could work as well)

4. For UDR when you add a svc=LoadBalancer for inbound traffic you need to think about the entire packet flow. What we do for firewall is add a public IP on the firewall for each svc=LoadBalancer, then when traffic hits the firewall we redirect it to the SLB public IP for ingress and then on the way out (packet response) our UDR picks it up and pushes it back to the firewall. 

UDR is a very advanced networking scenario that requires you to setup egress yourself - it is not a quick way to avoid a public IP.

I am most curious about (1) currently. Where are you pointing the 0.0.0.0/0 traffic to in your route table? Can you see whatever that device is and confirm that it is receiving traffic?

**Comment by baboune** [comment]
Maybe it is a terminology issue, what is meant by UDR?

1. Yes, I would expect the same.
2. Correct. We can not use that due to Azure Policies set by our organization.
3. That is the case we are interested in.

> A route table associated to your custom subnet

We added that. 

> That route table to direct all egress traffic 0.0.0.0/0 to your own egress device (we use a firewall in the docs but other personal devices could work as well)

That is where I am unclear, what is the "own egress device" for? Or in short, since I have set the routes in the table, and have egress set to "0.0.0.0/0" pointing to "internet"., I would expect egress to work.

4. Yes. makes sense.

> Where are you pointing the 0.0.0.0/0 traffic to in your route table? Can you see whatever that device is and confirm that it is receiving traffic?

The 0.0.0.0/0 route is pointing to "internet".

**Comment by abarqawi** [comment]
@gtracer customer opened case with support , the main question here can he use UDR without AzureFirewall/Appliance NVA ?
he just want to bypass traffic using rout-table to internet without public IP

**Comment by gtracer** [comment]
@abarqawi I am not sure if you can. Would be best to loop in the route table team and see what they say.  It is out of AKS's UDR scope (we just ensure that packet goes to whatever device you point to). 

But my hypothesis is that you cannot use a route of "internet" without a NVA since the RT itself doesn't NAT packet's. So you can send the packet to the internet using some azure backbone but you won't be able to properly send the response back to the node/pod that made the request because there will be no recoded pathway  

Please lmk what the RT team says as well / loop me into that thread

**Comment by baboune** [comment]
> AKS's UDR scope (we just ensure that packet goes to whatever device you point to).

@gtracer Where is that specified in the Azure documentation that a "device" is required?  

From the service description, it seems to only require "UDR" ie "user defined routes" which IMHO means a route table and rules.  Hence please clarify the feature and its associated documentation as a start as this requirement is missing.

**Comment by palma21** [comment]
It doesn't need a device but in that case how can traffic come back to the cluster if the IP that it leaves with is a private IP that is non-routable?

Loosely defined there are 2 types of egress in Azure:
* Explicit (eg. you have an outbound pool on an SLB or a UDR pointing to a FW that has a Public IP, or you have a Public IP on the node)
* Implicit (this is the case where you create an IaaS standalone VM in azure without public IP, in this case a public IP from a pool of azure managed IPs is selected at random for that VM to egress from), this is also the case with BLB – based architecture


Using SLB architecture you need to use explicit egress, so you need to send to a FW, GW, on-prem or allow the egress to be done by a PIP on SLB or the node.

If you want implicit egress you can use Basic Load Balancer that will give you that.

`az aks create -n aks -g aks --load-balancer-sku basic`

**Comment by baboune** [comment]
@palma21 Could you acknowledge the poor documentation and implementation of UDR?  It is rather uncivil to close an issue like that. And in particular with an argument that is that it does require a device.

Reverse order.

The Basic Load Balancer does not support multiple node pools nor AZ.  And AKS barely supports AZ as it is.  BLB is not an option.

> It doesn't need a device but in that case how can traffic come back to the cluster if the IP that it leaves with is a private IP that is non-routable?

Considering that the whole cluster is functional until the SLB is activated (ie until a Service is using type LoadBalancer is instantiated), could it be expected that it would still work then?  

The fact that SLB instance stops routing egress is internal Azure routing.

In short what is called the "implicit" failure of SLB is undocumented same as the UDR requirements that a device is specified.

AWS and GCP dont have those same limitations.  And those features are well documented.

From a customer service perspective, I am unimpressed by Azure.  Closing a ticket like that with such a poor answer.  Poor form and content.

**Comment by palma21** [comment]
@baboune i mentioned it _doesn't_ need a device. That is not the the problem but the fact you need a public IP. I'm happy to change any docs or contact the correct teams that maintain those docs if you believe they need clarification (since you rightly pointed they don't mention you need a device, which you don't, they'd be correct from your standpoint, if that's not the case happy to revisit it.

> Considering that the whole cluster is functional until the SLB is activated (ie until a Service is using type LoadBalancer is instantiated), could it be expected that it would still work then?

Yes it would, unfortunately that is not an option we can easily support as of today because nothing would stop a user from going ahead and creating an service which would incapacitate the whole cluster. We are working on this scenario but it requires a large set of validations and change in k8s upstream cloud provider for  this to work.

If your ask is for AKS to support this, I can re-open this as a feature request


> In short what is called the "implicit" failure of SLB is undocumented same as the UDR requirements that a device is specified.

We tried to document it here in the limitations of SLB, we can make the verbiage more clear. 

https://docs.microsoft.com/en-us/azure/aks/load-balancer-standard#limitations

I want to clarify that the GH closure was a misunderstanding and an error on my part since this was being tracked as clarification questions which wouldn't require the ticket to be open for the discussion to continue, I see that you actually have feature and doc asks so I am reopening until those are met. If you could clarify my above statements we can get that started.

**Comment by baboune** [comment]
To clarify, we must use UDR is because our internal Enterprise IT adds a policy violation on using any Public IP(PIP) before we receive the Azure subscription.

> https://docs.microsoft.com/en-us/azure/aks/load-balancer-standard#limitations

From my perspective, the SLB linked documentation refers to the general SLB ie the one using PIP while UDR tends to presented as a SLB w/o PIP e.g.  https://docs.microsoft.com/en-us/azure/aks/egress-outboundtype:

> However, the default setup may not meet the requirements of all scenarios if public IPs are disallowed or additional hops are required for egress.

Referring to https://docs.microsoft.com/en-us/azure/aks/egress-outboundtype#limitations or https://docs.microsoft.com/en-us/azure/aks/egress-outboundtype#outbound-type-of-userdefinedrouting, there are no clear mention of those limitations.

Overall, I would say that it is difficult to understand that a PiP is required.  In particular when from an egress perspective one could expect any of the VNET IPs to be implicitly NATed.

One documentatiom improvement could be to update the statement (part of https://docs.microsoft.com/en-us/azure/aks/egress-outboundtype#outbound-type-of-userdefinedrouting):

> A valid user-defined route (UDR) must exist on the subnet with outbound connectivity."

To include what you mentioned above ie:

> Using SLB architecture you need to use explicit egress, so you need to send to a FW, GW, on-prem or allow the egress to be done by a PIP on SLB or the node.

While I understand that this is how it is, and since we can not instantiate anything w a PIP, that means we can create an AKS + SLB cluster but not route traffic to it. OK.

In general, I would appreciate more in-depth articles describing how Load Balancer and the Azure CNI work as, from my perspective, the current requirements/limitations are unclear.  

Yes, I would like that allowing AKS + SLB w/o PIP to be re-opened as a feature request.

> clarification questions which wouldn't require the ticket to be open

In that case, not being aware of the process, I am sorry for the reaction.

**Comment by palma21** [comment]
Makes sense

@mlearned could we add this to the doc asap please:

> One documentatiom improvement could be to update the statement (part of https://docs.microsoft.com/en-us/azure/aks/egress-outboundtype#outbound-type-of-userdefinedrouting):

>>A valid user-defined route (UDR) must exist on the subnet with outbound connectivity."

>To include what you mentioned above ie:

>>Using SLB architecture you need to use explicit egress, so you need to send to a FW, GW, on-prem or allow the egress to be done by a PIP on SLB or the node.

I will leave this open for the feature request and tag it accordingly. (If you don't mind I'll update the title to reflect it)

**Comment by baboune** [comment]
Thank you.

**Comment by KylePolansky** [comment]
I just frustratingly spent the past week fighting with this issue. When reading the documentation, I incorrectly assumed that

> When using standard load balancer (SLB) architecture you must establish explicit egress. This requires sending egress requests to an appliance such as a firewall, gateway, on-prem or to allow the egress to be done by a public IP assigned to the standard load balancer or a given node.

would mean that a 0.0.0.0/0 route to the internet would be sufficient egress.

The default routes are working for us with basic load balancer and I didn't realize that SLB would have all these additional restrictions. Perhaps I'm a bad reader assuming that the implicit NAT'ed public IP associated with the 0.0.0.0/0 route would be magically associated to the SLB like it is appears to be in basic load balancer, but it might be worth explicitly adding to the docs that a 0.0.0.0/0 internet route will not work to make this restriction more clear.

I also noticed that if you don't have a 0.0.0.0/0 route in the subnet when creating a cluster with userDefinedRouting (as in my first attempt at UDR), the pre-creation validation will fail. Perhaps this check could be amended to fail if the 0.0.0.0/0 route is pointed to internet until this issue is resolved.

As OP mentioned, it's not clear at all why the cluster seemingly randomly fails when an internal load balancer is added. I didn't realize the problem was with internal load balancer for a very long time. I spun up countless vnets, route tables, role assignments, clusters, kubernetes templates, etc. until I was finally able to isolate the problem.

**Background on my issue:** I have similar restrictions to OP (no public IPs).

We have an existing kubenet cluster that's attached to a kubernetes managed subnet with a basic load balancer. This works great for us. We have all our internally exposed kubernetes resources in a single subnet without a public IP.

However, an upcoming application is too big to fit in our existing nodes (we need a node with more CPU resources). Because you can't change the size of the existing system pool and can't create another pool with basic load balancer, I'm forced to recreate the cluster. Because this is a rather long process (I'll have to schedule downtime, backup and restore all the data in our stateful pods, etc.) I've been looking at how we can prevent this problem in the future.

Since SLB supports multiple node pools, I thought simply changing from basic load balancer to standard load balancer would solve this problem. This would allow us to create new user pool for "special" applications like this, or resize the system pool without having to recreate the entire cluster. We are thinking about adding a windows node pool once kubenet is supported, and a SLB would be required for this as well.

My first attempt at SLB a while back created a public IP which was no good. Now that userDefinedRoute is supported in kubenet (https://github.com/Azure/AKS/blob/master/CHANGELOG.md#release-2020-06-01), I thought I'd give it another go with a very basic route table, but now ran into this issue.

**Repro steps for SBL:**
```powershell
# Setup: Create a Service Principal and paste the appId and secret below
# Setup: Create a new resource group and paste it's subscriptionid and name below

$servicePrincipalId = 'TODO'
$servicePrincipalSecret = 'TODO'
$subscriptionId = 'TODO'
$resourceGroup = 'TODO'


# Assign owner to resource group so we don't have to worry about permission assignments for this test
az role assignment create --assignee $servicePrincipalId --scope "/subscriptions/$subscriptionId/resourceGroups/$resourceGroup" --role Owner

# Create isolated vnet so we don't have to worry about other things in the vnet for this test
$vnetName = "$resourceGroup-vnet"
az network vnet create --name $vnetName --resource-group $resourceGroup --address-prefix 10.0.0.0/16 --subnet-name kubesubnet --subnet-prefix 10.0.0.0/24

# Create a default route table for the subnet that sends non-vnet egress to the internet
$routeTableName = "$resourceGroup-route-table"
az network route-table create --resource-group $resourceGroup --name $routeTableName
az network route-table route create --resource-group $resourceGroup --name default --route-table-name $routeTableName --address-prefix 0.0.0.0/0 --next-hop-type Internet

# Add route table to vnet
az network vnet subnet update --resource-group $resourceGroup --vnet-name $vnetName --name kubesubnet --route-table $routeTableName

# Create new aks cluster attached to the new subnet we just created
# Note: Some settings like the auto scaler and internal ip ranges are from a previous test and should be inconsequential in this example
$aksName = "$resourceGroup-kube"
az aks create `
    --resource-group $resourceGroup `
    --name $aksName `
    --enable-cluster-autoscaler `
    --node-count 1 `
    --min-count 1 `
    --max-count 2 `
    --node-vm-size Standard_D2s_v3 `
    --kubernetes-version 1.16.9 `
    --load-balancer-sku standard `
    --network-plugin kubenet `
    --outbound-type userDefinedRouting `
    --no-ssh-key `
    --service-cidr 10.72.135.0/24 `
    --dns-service-ip 10.72.135.10 `
    --pod-cidr 10.72.192.0/18 `
    --docker-bridge-address 172.17.0.1/16 `
    --vnet-subnet-id "/subscriptions/$subscriptionId/resourceGroups/$resourceGroup/providers/Microsoft.Network/virtualNetworks/$vnetName/subnets/kubesubnet" `
    --service-principal $servicePrincipalId `
    --client-secret $servicePrincipalSecret
```

Then try to deploy a service with an internal load balancer. For example, `kubectl apply -f my-nginx.yaml` where my-nginx.yaml is:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-nginx
spec:
  selector:
    matchLabels:
      run: my-nginx
  replicas: 1
  template:
    metadata:
      labels:
        run: my-nginx
    spec:
      containers:
      - name: my-nginx
        image: nginx
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: my-nginx
  labels:
    run: my-nginx
  annotations:
    service.beta.kubernetes.io/azure-load-balancer-internal: "true"
spec:
  type: LoadBalancer
  ports:
  - port: 80
    protocol: TCP
  selector:
    run: my-nginx
```

A working example with basic load balancer (our existing setup):
```powershell
# Setup: Create a Service Principal and paste the appId and secret below
# Setup: Create a new resource group and paste it's subscriptionid and name below

$servicePrincipalId = 'TODO'
$servicePrincipalSecret = 'TODO'
$subscriptionId = 'TODO'
$resourceGroup = 'TODO'


# Assign owner to resource group so we don't have to worry about permission assignments for this test
az role assignment create --assignee $servicePrincipalId --scope "/subscriptions/$subscriptionId/resourceGroups/$resourceGroup" --role Owner

# Create isolated vnet so we don't have to worry about other things in the vnet for this test
$vnetName = "$resourceGroup-vnet"
az network vnet create --name $vnetName --resource-group $resourceGroup --address-prefix 10.0.0.0/16 --subnet-name kubesubnet --subnet-prefix 10.0.0.0/24

# Create new aks cluster attached to the new subnet we just created
# Note: Some settings like the auto scaler and internal ip ranges are from a previous test and should be inconsequential in this example
$aksName = "$resourceGroup-kube"
az aks create `
    --resource-group $resourceGroup `
    --name $aksName `
    --enable-cluster-autoscaler `
    --node-count 1 `
    --min-count 1 `
    --max-count 2 `
    --node-vm-size Standard_D2s_v3 `
    --kubernetes-version 1.16.9 `
    --load-balancer-sku basic `
    --network-plugin kubenet `
    --no-ssh-key `
    --service-cidr 10.72.135.0/24 `
    --dns-service-ip 10.72.135.10 `
    --pod-cidr 10.72.192.0/18 `
    --docker-bridge-address 172.17.0.1/16 `
    --vnet-subnet-id "/subscriptions/$subscriptionId/resourceGroups/$resourceGroup/providers/Microsoft.Network/virtualNetworks/$vnetName/subnets/kubesubnet" `
    --service-principal $servicePrincipalId `
    --client-secret $servicePrincipalSecret
```

**Conclusion:**
The only changes from our current template are changing the load balancer from basic to standard and going from managed routing to user defined routing to avoid the public IP.

However, SLB seems to have some restrictions that are preventing this from happening. I think ideally I'd be able to apply SLB with kubernetes managed subnet routing, but that does not seem currently possible without a public IP.

**Comment by palma21** [comment]
BLB also has a public IP involved, you just don't see it. UDR without an actual UDR (or a UDR to the Internet) is not possible, in fact UDR to internet is already part of the default UDRs in a route table, but again without a public IP there's no way to SNAT to the internet.

There's two things to call out here:
1) AKS requires some outbound access for things like MCR (Container Registry) and packages.microsoft.com so that the nodes can bootstrap and continue to work normally over time.
2) For outbound access you need a Public IP somewhere to SNAT.
   - For BLB that IP to SNAT is a random Public IP on the Azure infra (causes issues later on around SNAT exhaustion, and BLB itself can't load balance across multiple node pools, etc.)
   - As such AKS moved to SLB. SLB has explicit egress which means you need to define the public IP to SNAT, and at the same time gives you more flexibility to have multiple outbound IPs and outbound ports.
   - UDR is the model for folks that can't egress or create public IPs in their AKS envs. To use it you need to provide the UDR to whatever NVA, proxy, firewall or GW your architecture uses to outbound. (again the outbound path is required)

We've update the docs to container this:
https://docs.microsoft.com/en-us/azure/aks/egress-outboundtype#deploy-a-cluster-with-outbound-type-of-udr-and-azure-firewall

Today for UDR the SLB IP and the SLB itself is no longer created. 

The ask for SLB + Implicit is not possible as per the above discussions we have feature requests open for:

- Managed NAT support (public IP on the Azure managed NAT service) - https://github.com/Azure/AKS/issues/1122
- AKS no-outbound (this is currently a feature ask to the MCR team to private private connections to MCR). Once that is done we can complete a feature that allows for AKS clusters to not require outbound connection, and as such no need for NAT'ing.

As such I'm closing this issue, as the requests have been broken down, docs have been updated and answers provided.
Feel free to comment if you need additional clarifications.
